File: Sources/Sampler/metropolis_hastings_pt.cc
// Copyright 2019 The Simons Foundation, Inc. - All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "Sampler/metropolis_hastings_pt.hpp"

namespace netket {

MetropolisHastingsPt::MetropolisHastingsPt(
    AbstractMachine& machine,
    MetropolisHastingsPt::TransitionKernel transition_kernel, Index n_replicas,
    Index sweep_size)
    : AbstractSampler(machine),
      transition_kernel_(transition_kernel),
      n_replicas_(n_replicas),
      sweep_size_(sweep_size) {
  detail::CheckReplicasSize(__FUNCTION__, n_replicas);
  detail::CheckSweepSize(__FUNCTION__, sweep_size);

  current_X_.resize(n_replicas, machine.Nvisible());
  proposed_X_.resize(n_replicas, machine.Nvisible());
  current_Y_.resize(n_replicas);
  proposed_Y_.resize(n_replicas);
  quotient_Y_.resize(n_replicas);
  probability_.resize(n_replicas);
  accept_.resize(n_replicas);
  log_acceptance_correction_.resize(n_replicas);

  beta_.resize(n_replicas);
  proposed_beta_.resize(n_replicas);

  accepted_samples_.resize(n_replicas);

  // Linearly spaced inverse temperature
  for (Index i = 0; i < n_replicas; i++) {
    beta_(i) = (1. - double(i) / double(n_replicas));
  }

  beta1_ind_ = 0;

  Reset(true);
}

void MetropolisHastingsPt::Reset(bool init_random) {
  if (init_random) {
    for (Index i = 0; i < n_replicas_; i++) {
      GetMachine().GetHilbert().RandomVals(current_X_.row(i),
                                           GetRandomEngine());
    }
  }
  GetMachine().LogVal(current_X_, current_Y_, {});

  accepted_samples_.setZero();
  total_samples_ = 0;
  total_exchange_steps_ = 0;
  beta1_av_ = 0;
  beta1_av_sq_ = 0;
}

std::pair<Eigen::Ref<const RowMatrix<double>>,
          Eigen::Ref<const Eigen::VectorXcd>>
MetropolisHastingsPt::CurrentState() const {
  return {Eigen::Map<const RowMatrix<double>>{current_X_.row(beta1_ind_).data(),
                                              1, current_X_.cols()},
          Eigen::Map<const Eigen::VectorXcd>{&current_Y_(beta1_ind_), 1}};
}

void MetropolisHastingsPt::SetVisible(Eigen::Ref<const RowMatrix<double>> x) {
  CheckShape(__FUNCTION__, "v", {x.rows(), x.cols()}, {1, current_X_.cols()});

  current_X_.row(beta1_ind_) = x.row(0);
}

void MetropolisHastingsPt::OneStep() {
  transition_kernel_(current_X_, proposed_X_,
                     log_acceptance_correction_);  // Now proposed_X_ contains
                                                   // next states `v'`

  GetMachine().LogVal(proposed_X_, /*out=*/proposed_Y_, /*cache=*/{});

  // Calculates acceptance probability
  quotient_Y_ =
      ((proposed_Y_ - current_Y_ + log_acceptance_correction_) * beta_);

  probability_ = (GetMachinePow() * quotient_Y_.real()).exp();

  for (auto i = Index{0}; i < accept_.size(); ++i) {
    accept_(i) = probability_(i) >= 1.0
                     ? true
                     : std::uniform_real_distribution<double>{}(
                           GetRandomEngine()) < probability_(i);

    // Updates current state
    if (accept_(i)) {
      current_X_.row(i) = proposed_X_.row(i);
      // Update acceptance counters
      accepted_samples_(i) += 1.0;
    }
  }

  current_Y_ = accept_.select(proposed_Y_, current_Y_);

  total_samples_ += 1;
}

void MetropolisHastingsPt::ExchangeStep() {
  // Choose a random swap order (odd/even swap)
  Index swap_order =
      std::uniform_int_distribution<Index>(0, 1)(GetRandomEngine());

  ProposePairwiseSwap(beta_, proposed_beta_, swap_order);

  quotient_Y_ = ((proposed_beta_ - beta_) * current_Y_);
  probability_ = (GetMachinePow() * quotient_Y_.real()).exp();

  for (auto i = Index{swap_order}; i < probability_.size(); i += 2) {
    Index inn = (i + 1) % n_replicas_;

    probability_(i) *= probability_(inn);
    accept_(i) = probability_(i) >= 1.0
                     ? true
                     : std::uniform_real_distribution<double>{}(
                           GetRandomEngine()) < probability_(i);

    // Updates current state
    if (accept_(i)) {
      std::swap(beta_(i), beta_(inn));
      std::swap(accepted_samples_(i), accepted_samples_(inn));

      // Keep track of the position of beta=1
      if (beta1_ind_ == i) {
        beta1_ind_ = inn;
      } else if (beta1_ind_ == inn) {
        beta1_ind_ = i;
      }
    }
  }

  total_exchange_steps_ += 1.0;

  // Update statistics to compute diffusion coefficient of replicas
  double delta = beta1_ind_ - beta1_av_;
  beta1_av_ += delta / total_exchange_steps_;
  double delta2 = beta1_ind_ - beta1_av_;
  beta1_av_sq_ += delta * delta2;
}

Index MetropolisHastingsPt::BatchSize() const noexcept { return n_replicas_; }
Index MetropolisHastingsPt::NChains() const noexcept { return 1; }
Index MetropolisHastingsPt::SweepSize() const noexcept { return sweep_size_; }

void MetropolisHastingsPt::SweepSize(Index const sweep_size) {
  detail::CheckSweepSize(__FUNCTION__, sweep_size);
  sweep_size_ = sweep_size;
}

void MetropolisHastingsPt::Sweep() {
  for (auto i = Index{0}; i < sweep_size_; ++i) {
    OneStep();
    ExchangeStep();
  }
}

void MetropolisHastingsPt::ProposePairwiseSwap(
    Eigen::Ref<const Eigen::ArrayXd> vin, Eigen::Ref<Eigen::ArrayXd> vout,
    int order) {
  assert(order == 0 || order == 1);
  assert(vin.size() == vout.size());
  assert(vin.size() % 2 == 0);

  if (order == 0) {
    for (int i = 0; i < vin.size(); i += 2) {
      vout(i) = vin(i + 1);
      vout(i + 1) = vin(i);
    }
  } else if (order == 1) {
    for (int i = 1; i < vin.size() - 1; i += 2) {
      vout(i) = vin(i + 1);
      vout(i + 1) = vin(i);
    }
    vout(0) = vin(vin.size() - 1);
    vout(vout.size() - 1) = vin(0);
  }
}

std::map<std::string, double> MetropolisHastingsPt::Stats() const {
  NETKET_CHECK(total_samples_ > 0 && total_exchange_steps_ > 0, RuntimeError,
               "Cannot compute sampler stats, because no moves were made");

  std::map<std::string, double> stats;

  Eigen::ArrayXd accept =
      accepted_samples_ / static_cast<double>(total_samples_);

  stats["mean_acceptance"] = accept.mean();
  stats["min_acceptance"] = accept.minCoeff();
  stats["max_acceptance"] = accept.maxCoeff();

  // Average position of beta=1
  // This is normalized and centered around zero
  // In the ideal case the average should be zero
  stats["normalized_beta=1_position"] =
      beta1_av_ / static_cast<double>(n_replicas_ - 1) - 0.5;

  // Average variance on the position of beta=1
  // In the ideal case this quantity should be of order ~ [0.2, 1]
  stats["normalized_beta=1_diffusion"] =
      std::sqrt(beta1_av_sq_ / total_exchange_steps_) /
      static_cast<double>(n_replicas_);
  return stats;
}

}  // namespace netket

File: netket/legacy/sampler/numpy/metropolis_hastings_pt.py
import numpy as _np
from ..abstract_sampler import AbstractSampler

from netket.legacy.stats import mean as _mean

from numba import jit
from numba import int64, float64
from netket.legacy import random as _random
import math


class MetropolisHastingsPt(AbstractSampler):
    def __init__(
        self,
        machine,
        kernel,
        n_replicas=32,
        sweep_size=None,
    ):
        super().__init__(machine)

        self.n_replicas = n_replicas

        self.sweep_size = sweep_size

        self._kernel = kernel

        self.machine_pow = 2.0

        # Contains quantities to compute diffusion coefficient of replicas
        # beta_stats[0] = position of replica at beta=1 (do not reset this!)
        # beta_stats[1] = running average of beta=1 position
        # beta_stats[2] = running average of beta=1 position**2
        # beta_stats[3] = current number of exchange steps performed
        self._beta_stats = _np.zeros(4)

        self.reset(True)

    @property
    def n_replicas(self):
        return self._n_replicas

    @n_replicas.setter
    def n_replicas(self, n_replicas):
        if n_replicas < 1:
            raise ValueError("Expected n_replicas>0. ")

        self._n_replicas = n_replicas

        self._state = _np.zeros((n_replicas, self._input_size))
        self._state1 = _np.copy(self._state)

        self._log_values = _np.zeros(n_replicas, dtype=_np.complex128)
        self._log_values_1 = _np.zeros(n_replicas, dtype=_np.complex128)
        self._log_prob_corr = _np.zeros(n_replicas)

        # Linearly spaced inverse temperature
        self._beta = _np.empty(n_replicas)

        for i in range(n_replicas):
            self._beta[i] = 1.0 - float(i) / float(n_replicas)

        # some temporary arrays
        self._proposed_beta = _np.empty(n_replicas)
        self._beta_prob = _np.empty(n_replicas)

    @property
    def machine_pow(self):
        return self._machine_pow

    @machine_pow.setter
    def machine_pow(self, m_power):
        self._machine_pow = m_power

    @property
    def sweep_size(self):
        return self._sweep_size

    @sweep_size.setter
    def sweep_size(self, sweep_size):
        self._sweep_size = sweep_size if sweep_size != None else self._input_size
        if self._sweep_size < 0:
            raise ValueError("Expected a positive integer for sweep_size ")

    def reset(self, init_random=False):
        if init_random:
            self._kernel.random_state(self._state)

        self._log_values = self.machine.log_val(self._state, out=self._log_values)

        self._accepted_samples = _np.zeros(self._n_replicas)
        self._total_samples = 0

        self._beta_stats[1:].fill(0)

    @staticmethod
    @jit(nopython=True)
    def _fixed_beta_acceptance_kernel(
        state,
        state1,
        log_values,
        log_values_1,
        log_prob_corr,
        machine_pow,
        beta,
        accepted,
    ):

        for i in range(state.shape[0]):
            prob = math.exp(
                machine_pow
                * beta[i]
                * (log_values_1[i] - log_values[i] + log_prob_corr[i]).real
            )

            assert not math.isnan(prob)

            accept = prob > _random.uniform(0, 1)

            if accept:
                log_values[i] = log_values_1[i]
                state[i] = state1[i]
                accepted[i] += 1

    def __next__(self):

        _log_val = self.machine.log_val
        _fb_acc_kernel = self._fixed_beta_acceptance_kernel
        _state = self._state
        _state1 = self._state1
        _log_values = self._log_values
        _log_values_1 = self._log_values_1
        _log_prob_corr = self._log_prob_corr
        _machine_pow = self._machine_pow
        _accepted_samples = self._accepted_samples
        _t_kernel = self._kernel.transition
        _beta = self._beta
        _beta_stats = self._beta_stats
        _proposed_beta = self._proposed_beta
        _beta_prob = self._beta_prob

        for sweep in range(self.sweep_size):

            # Propose a new state using the transition kernel
            _t_kernel(_state, _state1, _log_prob_corr)

            _log_values_1 = _log_val(_state1, out=_log_values_1)

            # Acceptance Kernel for fixed-beta moves
            _fb_acc_kernel(
                _state,
                _state1,
                _log_values,
                _log_values_1,
                _log_prob_corr,
                _machine_pow,
                _beta,
                _accepted_samples,
            )

            # Transition + Acceptance Kernel for replica exchange moves
            self._exchange_step_kernel(
                _log_values,
                _machine_pow,
                _beta,
                _proposed_beta,
                _beta_prob,
                _beta_stats,
                _accepted_samples,
            )

        self._total_samples += self.sweep_size
        return self._state[_np.intp(_beta_stats[0])].reshape(1, -1)

    @staticmethod
    @jit(nopython=True)
    def _exchange_step_kernel(
        log_values, machine_pow, beta, proposed_beta, prob, beta_stats, accepted_samples
    ):
        # Choose a random swap order (odd/even swap)
        swap_order = _random.randint(0, 2, size=()).item()

        n_replicas = beta.shape[0]

        for i in range(swap_order, n_replicas, 2):
            inn = (i + 1) % n_replicas
            proposed_beta[i] = beta[inn]
            proposed_beta[inn] = beta[i]

        for i in range(n_replicas):
            prob[i] = math.exp(
                machine_pow * (proposed_beta[i] - beta[i]) * log_values[i].real
            )

        for i in range(swap_order, n_replicas, 2):
            inn = (i + 1) % n_replicas

            prob[i] *= prob[inn]

            if prob[i] > _random.uniform(0, 1):
                # swapping status
                beta[i], beta[inn] = beta[inn], beta[i]
                accepted_samples[i], accepted_samples[inn] = (
                    accepted_samples[inn],
                    accepted_samples[i],
                )

                if beta_stats[0] == i:
                    beta_stats[0] = inn
                elif beta_stats[0] == inn:
                    beta_stats[0] = i

        # Update statistics to compute diffusion coefficient of replicas
        # Total exchange steps performed
        beta_stats[-1] += 1

        delta = beta_stats[0] - beta_stats[1]
        beta_stats[1] += delta / float(beta_stats[-1])
        delta2 = beta_stats[0] - beta_stats[1]
        beta_stats[2] += delta * delta2

    @property
    def stats(self):
        stats = {}
        accept = self._accepted_samples / float(self._total_samples)

        stats["mean_acceptance"] = _mean(accept)
        stats["min_acceptance"] = _mean(accept.min())
        stats["max_acceptance"] = _mean(accept.max())

        # Average position of beta=1
        # This is normalized and centered around zero
        # In the ideal case the average should be zero
        stats["normalized_beta=1_position"] = (
            self._beta_stats[1] / float(self._n_replicas - 1) - 0.5
        )

        # Average variance on the position of beta=1
        # In the ideal case this quantity should be of order ~ [0.2, 1]
        stats["normalized_beta=1_diffusion"] = _np.sqrt(
            self._beta_stats[2] / self._beta_stats[-1]
        ) / float(self._n_replicas)

        return stats
